### ISSUE TITLE
Shorthand for "TypedGlamor.toString"?

### ISSUE THREAD

**Issue by emmenko** [glennsl/bs-typed-glamor#3]
Another nitpick: I found myself writing `TypedGlamor.toString` every time I need to pass a class to a React element. Even though I `open TypedGlamor;` on top of the module, I find it a bit confusing of simply writing `toString`.

Therefore, I was wondering if we can provide another alias or something to make this shorter, for example `cs` (for _class_ or _class string_), so you can write `Styles.header |> cs`.

WDYT?

**Comment by glennsl** [comment]
I definitely think there should be an easier way to consume styles, but I don't think unclear shorthands such as `cs` are good in general. It's also not a good idea to `open TypedGlamor` at the top of any module other than those dealing solely with CSS since it heavily pollutes the namespace. Module-level open's are generally discouraged in OCaml (unless the modules opened are specifically designed for it) because it causes significant problems.

I've been trying to think of more usable ways to attach styles to components, like a HOC or function to wrap the component perhaps, but haven't come across anything that really hits the spot yet. Ideas are certainly welcome.

**Comment by emmenko** [comment]
> It's also not a good idea to open TypedGlamor at the top of any module other than those dealing solely with CSS since it heavily pollutes the namespace

Hmm that's true. I actually have a `module Styles` definition inside the file, so I ~~can~~ should just `open` inside there. Thanks for the reminder!

> I've been trying to think of more usable ways to attach styles to components, like a HOC or function to wrap the component perhaps, but haven't come across anything that really hits the spot yet. Ideas are certainly welcome.

Hmm how about wrapping the `css` constructor when you define the styles, so that the variable value is already a string. For example:

```reason
module Styles = {
  open TypedGlamor;

  let header = cssAsString([
    color(hex("ccc")),
  ]);
};

<div className=Styles.header />
```

Where `cssAsString` simply does `css(...) |> toString`.

**Comment by glennsl** [comment]
Then you can't extend it. It's definitely a possibility, but not something I want to commit to yet.

**Comment by glennsl** [comment]
A compromise might be to use a lookup table to find the style object from the class name, which can then be used when extending. But that has a performance impact that might be too great. Something to consider though.

**Comment by emmenko** [comment]
> Then you can't extend it.

What do you mean by "extend"? (I'm not yet familiar with the API...)

**Comment by glennsl** [comment]
Extending means reusing and adding to a previously defined style. `css` has an optional `extend` argument: https://github.com/glennsl/bs-typed-glamor/blob/master/src/TypedGlamor.re#L959-L966

You can see it in action here: https://github.com/redex/redex.github.io/blob/source/src/styles/PackageStyles.ml#L144

This API also isn't set in stone of course

**Comment by emmenko** [comment]
I see. I don't see much a problem though, the `cssAsString` has the same signature of the `css` function (so you can still use extend). Obviously if you know that a style should be used to extend other styles, you just use `css`.
The `cssAsString` is just syntactic sugar to do the string conversion.

```reason
module Styles = {
  open TypedGlamor;

  let base = css([color(red)]);

  let header = cssAsString(
    ~extend=base,
    [color(hex("ccc"))]
  );
};

<div className=Styles.header />
```

So `base` returns a declaration and is used to extend `header`. However, `header` (which is the "final" style) already returns a string, so I can just use the variable.

Hope that makes sense where I wanted to get with this.

Of course you have much more experience with those sorts of APIs, maybe my proposal just doesn't make sense.

**Comment by emmenko** [comment]
It doesn't need to be a new function, it can also be a new argument in the `css` method. For example:

```reason
let css = (~extend=?, ~toString=false, declarations) => {
  let this = declarations |> Declarations.toDict;

  let s = switch extend {
  | Some(style) => merge([|style, css(this)|])
  | None => css(this)
  };

  toString ? Js.String.make(s) : s;
};
```

**Comment by glennsl** [comment]
The problem is that it is a messy API, where you have to keep track of what can and cannot be extended, what can and cannot be used directly, and if you want to switch from one to the other you'll have to make changes in lots of different places. It's also possible to accomplish the same now by just doing `css(...) |> toString;`. 

Ideally you wouldn't have to deal with converting to a string at all though, and I'd also like to be able to use hard-coded identifiers instead of strings in selectors to make that type-safe (i.e. `selectDescendants(myClass, [...])` instead of `select("& > .myClass", [...]`), which will probably also affect this API.

**Comment by emmenko** [comment]
True. I mean it's perfectly fine as it is now and this was more of a nice-to-have question. I think we can focus on improving other parts first 😄

PS: regarding the selectors, in case it helps, I usually only need the [lobotomized owl](https://css-tricks.com/lobotomized-owls/) selector

```reason
select("> * + *", css([]))
```
